packages/grafana-runtime/src/utils/DataSourceWithBackend.ts

```diff
@@ -379,6 +379,34 @@ class DataSourceWithBackend<
       });
     });
   }
+
+  /**
+   * EXPERIMENTAL. Calls migration endpoint. Requires grafanaAPIServerWithExperimentalAPIs or datasourceAPIServers feature toggle.
+   */
+  async postMigrateQuery(query: TQuery): Promise<TQuery> {
+    if (!(config.featureToggles.grafanaAPIServerWithExperimentalAPIs || config.featureToggles.datasourceAPIServers)) {
+      console.warn('migrateQuery is only available with the experimental API server');
+      return query;
+    }
+    const { apiVersion, type } = this.getRef();
+    if (!apiVersion || !type) {
+      console.warn('migrateQuery called with datasource that does not have apiVersion, type or uid set');
+      return query;
+    }
+
+    const dsnameURL = type.replace(/^(grafana-)?(.*?)(-datasource)?$/, '$2');
+    const url = `/apis/${dsnameURL}.datasource.grafana.app/${apiVersion}/namespaces/${config.namespace}/queryconvert`;
```

Comment:
I think we need to revisit how the frontend plugin knows its group name.  I think we should:
1. include the *group* in the DTO returned from: https://play.grafana.org/api/plugins
2. the version must be hardcoded in the client datasource.  It must not try to guess what version it knows how to use otherwise we are defeating the purpose of versioned APIs

Reply:
> 1\. include the _group_ in the DTO returned from: https://play.grafana.org/api/plugins

mmh, I am not sure about modifying the "legacy" api with things only for /apis. What if https://play.grafana.org/apis returned the plugin id for the datasource groups? (not sure if that's possible).

> 2\. the version must be hardcoded in the client datasource.  It must not try to guess what version it knows how to use otherwise we are defeating the purpose of versioned APIs

Not sure I understand. Thinking about this, it probably has a bug. Right now it's using the API version of the stored datasource (this is the API version of the plugin for whenever the datasource was saved). In my head, it should use the API version of the plugin whenever the _query_ was saved but if I understand you correctly, you want to use the API version that is currently running (the preferred version returned by calling /apis rather than having to hardcode it?). This means that we could potentially send a v1 query to a v2 backend. Is that the expected behavior?

Reply:
re including group in the legacy api only helps with using "v0" -- where we expect to be able to send the same payloads we currently send.

Once we have any explicit version -- the client *must* know that it want to talk to v1 or v2.    I think there is logic somehere that uses the preferredVersion -- this is for sure a bug (or able to introduce bugs) since the client needs to know what api version it understands

Reply:
> What if https://play.grafana.org/apis returned the plugin id for the datasource groups? (not sure if that's possible).

We will likely want something like this -- there will be a resource that represents all plugins/apps that we can inspect.  For now I don't see any harm adding it to the dto just to simplify transition and not get blocked till we have a complete story for all plugins as apisevers

Reply:
> the client _must_ know that it want to talk to v1 or v2. I think there is logic somehere that uses the preferredVersion -- this is for sure a bug (or able to introduce bugs) since the client needs to know what api version it understands

You are talking about [this](https://github.com/grafana/grafana/blob/postMigrateQuery/public/app/features/apiserver/client.ts#L118), right? We have discussed this topic before and I thought we concluded that approach (calling /apis to retrieve a preferredVersion) was correct. The client (Grafana UI) is retrieving the plugin code (including the QueryEditor) from the backend so I don't see why using also the version the backend returns is a bad idea. The other option is to inject the apiVersion in the plugin frontend at build time (which we already tried and removed from the plugin.json) but I don't see any major advantage of that.

> For now I don't see any harm adding it to the dto just to simplify transition and not get blocked till we have a complete story for all plugins as apisevers

Is there any interface we can use in the backend to retrieve the group info? Right now, the /api/plugins information is populated from the plugin.json info. Are you suggesting to use plugin.json for this or something else?

I think it's simpler to just mimic the logic from https://github.com/grafana/grafana/blob/convertObjectsQueries/pkg/plugins/apiserver.go#L10 for the time being, which is just droping the `grafana` org and the `datasource` type from the plugin ID.